circuits/circuits/distribute/README.md

```diff
@@ -0,0 +1,48 @@
+There is a `filtered_count` input that starts at `0`.
+
+The circuit processes one claim per iteration/fold. The claim accumulator starts at 0, one claim is added per iteration, and it must end with the right onchain value.
+If the claim exists in the state history, the `filtered_count` variable is incremented and the claim is added to a filtered poseidon merkle tree at position `filtered_count`. Otherwise, `filtered_count` stays the same and nothing is changed.
+
+If all the checks pass, then this circuit outputs a `filtered_tree` of size `filtered_count`, where every claim in the tree is valid.
+
+### Challenge Resolution
+
+The challenge resolution circuit is also in Nova and runs iteratively. The initial input is the `filtered_tree` which is a poseidon Merkle tree, and each iteration takes two contradictory claims in the tree, resolves the contradiction, and updates the tree.
+
+The first claim is called the winner and the second is the loser. The winner's block number must be greater than the loser's. The winner's leaf remains unedited. If the winner's coins are a superset of the loser's (i.e., $[0,5]$ beats $[2,3]$), the loser's leaf is deleted, and the `resolved_count` is decremented. If the loser's coins are split (i.e., $[2,3]$ beats $[0,5]$), the loser's leaf is deleted, but two new leaves are added and the `resolved_count` is incremented. If the loser's coins are truncated (i.e., $[1,5]$ vs $[1,3]$), the loser's leaf is deleted, a new leaf is added and the `resolved_count` remains the same.
+
+Now we have a `resolved_tree` that contains all valid winning claims, but may contain additional claims.
+
+### Proving Disjointness
+
+The disjointness circuit takes the `resolved_tree` as input, and a new `final_tree` as advice. Its goal is to show that the `final_tree` is sorted by coin order, fully disjoint, and only contains elements from the `resolved_tree`. This proves that the final tree only contains the true winners, and, importantly, it can be deterministically calculated with onchain data, so everyone can calculate the merkle proofs they need. The `final_tree` should use keccak for gas efficiency.
+
+The circuit iterates over every element in the `final_tree` and makes sure that its `end_coin` is lower than the `start_coin` of the next claim (unless it's the final claim, in which case there is no next claim). It also makes sure that the element exists in the `resolved_tree`.
+
+Since the `start_coin[i] < end_coin[i]` for every valid claim (if the protocol operated correctly while running!!), and `end_coin[i] < start_coin[i+1]` from our ordering check, we know that our claims are strictly ordered. This means that every claim is different. Since we have proved that the `final_tree` contains at least `resolve_count` distinct elements that also exist in the `resolved_tree`, we know that `final` is a superset of `resolved`. To prove that the sets are strictly equal, we have to prove that all other elements are zero.
```

Comment:
how have we proved this:
> we have proved that the `final_tree` contains at least `resolve_count` distinct elements that also exist in the `resolved_tree

Reply:
There are several parts to this:

1. Proving that `final_tree` contains elements from the `resolved_tree`
2. Proving that the elements in `final_tree` are distinct
3. Proving that there are at least `resolved_count` elements in `final_tree`

For `1.` we just use merkle proofs to show every element in `final_tree` is also in `resolved_tree`.
For `2.` we make sure that the coin ranges are always strictly increasing for each element in `final_tree`, so all elements must be unique.
I guess didn't really give a reason for `3.` actually! But we can do it by incrementing for each fold (similar to the other proofs) to count every element in `final_tree`, then just making sure that number is equal to `resolved_count`.

Reply:
I'll edit this to explain 3.